src/addons/ReactFragment.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright 2015, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+* @providesModule ReactFragment
+*/
+
+'use strict';
+
+var ReactElement = require('ReactElement');
+
+var warning = require('warning');
+
+/**
+ * We used to allow keyed objects to serve as a collection of ReactElements,
+ * or nested sets. This allowed us a way to explicitly key a set a fragment of
+ * components. This is now being replaced with an opaque data structure.
+ * The upgrade path is to call React.addons.createFragment({ key: value }) to
+ * create a keyed fragment. The resulting data structure is opaque, for now.
+ */
+
+if (__DEV__) {
+  var fragmentKey = '_reactFragment';
+  var canWarnForReactFragment = false;
+
+  try {
+    // Feature test. Don't even try to issue this warning if we can't use
+    // enumerable: false.
+
+    var dummy = function() {
+      return 1;
+    };
+
+    Object.defineProperty(
+      {},
+      fragmentKey,
+      { enumerable: false, value: true }
+    );
+
+    Object.defineProperty(
+      {},
+      'key',
+      { enumerable: true, get: dummy }
+    );
+
+    canWarnForReactFragment = true;
+  } catch (x) { }
+
+  var proxyPropertyAccessWithWarning = function(obj, key) {
+    Object.defineProperty(obj, key, {
+      enumerable: true,
+      get: function() {
+        warning(
+          false,
+          'A ReactFragment is an opaque type. You cannot access any of its ' +
```

Comment:
hm... Well technically you can because it is an upgrade path but we'll immediately break this next release so then you can't. Not sure how to phrase it to make it clear: Don't do that.